README.md
&nbsp;

&nbsp;

<div align="center"><a href="#"><img src="/images/colorparse-logo.png" alt="[;colorparse]" width=500 /></a>
</div>

<p align="center">
  Give <b><a href="#">color</a></b> and <b><i>style</i></b> to text effortlessly using commands!
</p>

<div align="center">

[![python version compatibility >= 3.7](https://img.shields.io/badge/python-%3E%3D%203.7-blue?style=flat-square&logo=python&logoColor=white&logoWidth=12)](https://pypi.org/project/colorparse/)&ensp;
[![package version 2.0.0](https://img.shields.io/pypi/v/colorparse?color=green&label=package&style=flat-square)](https://pypi.org/project/colorparse/)&ensp;
[![python wheel available](https://img.shields.io/pypi/wheel/colorparse?color=%23fca32f&style=flat-square)](https://pypi.org/project/colorparse/)&ensp;
[![documentation available](https://img.shields.io/badge/docs-yes-%23fc8a8a?style=flat-square)](/)&ensp;
[![changelog available](https://img.shields.io/badge/changelog-latest-%23ed63e3?style=flat-square)](/CHANGELOG.md)

</div>

&nbsp;

&nbsp;
  
## Table of contents

- [Table of contents](#table-of-contents)
- [Installation](#installation)
- [Usage](#usage)
  - [Concepts](#concepts)
  - [Examples](#examples)
  - [Module](#module)
    - [Methods](#methods)
- [Default commands](#default-commands)
- [See more](#see-more)
  - [License](#license)
  - [Full documentation](#full-documentation)
  - [More examples](#more-examples)
  - [Changelog](#changelog)
<h2></h2>
&nbsp;

&nbsp;
  
## Installation

```terminal
py -m pip install colorparse
```

<div style="height: 25px"></div>

## Usage

&nbsp;

**NOTE**: Because I believe this package might include a lot more bloat than neccessary for some cases, [colorparse-mini is also available]().

&nbsp;

The main difference between most text-coloring packages for python and *colorparse*, is that the former ones use imported variables that contain the colors. For example, you would use something like:

```python
>>> print(Colors.red + "roman empire") # With colorparse: 'colors.foreground.RED'
```

However, even though you are allowed to use this structure with this package (a little bit more verbose anyways), the expected way is using the following:

```python
>>> print("[;r]roman empire") # The "color command" is ;r for red
```

The idea is to avoid cluttering print statements or plain strings with concatenations, and instead use commands (customizable) which will later get parsed into colors or styles.

### Concepts

<details open>
  <summary><a name="commands"></a><b>Commands</b></summary>

A command is an object that represents a specific color or style and holds 3 components:
- A variable name (to store it and manually access through code).
- The proper command string (used to replace with the final color or style, like `;r` for red or `;+b`.
- The [ANSI escape code](https://en.wikipedia.org/wiki/ANSI_escape_code#SGR_(Select_Graphic_Rendition)_parameters) which is stored as a Code object inside the command.
  
&nbsp;
</details>
<details open>
  <summary><a name="types"></a><b>Types</b></summary>
  
For both colors and styles, there are two types. Colors have the types `foreground` (`;`) and `background` (`:`). Styles have `start` (`;+`) and `end` (`;-`).

This is useful when working around commands as variables (e.g. `colors.foreground.RED` or `styles.end.BOLD`), but also to disinguish between starting and ending styles.
  
&nbsp;
</details>
<details open>
  <summary><a name="stopping-commands"></a><b>Stopping commands</b></summary>
  
There are specific commands that stop foreground or background colors currently active: `;;` (any foreground color) and `::` (any background color). Use either `;:` or `:;` (interchangable) to stop any color regardless of type.

On the other hand, styles can be stopped using each style's counterpart (the `end` type). For example, if bold is used with `;+b` (with type `start`, because of the plus sign), the counterpart is `;-b` (of type `end`).

Use `:.` to stop all commands.
  
&nbsp;
</details>
<details open>
  <summary><a name="encapsulation"></a><b>Encapsulation</b></summary>
  
Any command can be encapsulated by surrounding it with special characters. This is useful to avoid situations where a command is next to text that is not meant to be part of the command like: "`;r`roman empire" (words "roman empire" in red) which would be interpreted as "`;rr`oman empire" ("oman empire" in **dark red** instead).

Commands can be surrounded by one or two forward slashes `/` on each side of the command, both regular parenthesis `(` and `)` or both square brackets `[` and `]`. Examples of these are: "`;r/`roman empire", "`(;r)`roman empire" and "`[;r]`roman empire". All result in the detection of red command, instead of the dark red variation.

***Note***: forward slashes can only be placed next to the command, unlike the parenthesis and square brackets which allow spaces in between.
  
&nbsp;
</details>
<details open>
  <summary><a name="escaping"></a><b>Escaping</b></summary>
  
If needed, commands can be escaped by using backslash `\` right before the start, not including encapsulation (e.g. "`\;rr`oman empire" returns ";rroman empire" and "`[\;r]`roman empire" returns "[;r]roman empire").

&nbsp;
</details>

&nbsp;

### Examples

```python
>>> import colorparse as cp # Use 'cp.print' to print and 'cp.paint' to return only
```

```python
>>> # ;g (green foreground)
>>> cp.print(";g/green goblin")
```

<div align="center" title='"green goblin" printed in green'>
  <img alt='"green goblin" printed in green' src="/images/colorparse-example-result-1-dark.png" width="288px" />
  <img src="/images/colorparse-example-result-1-light.png" width="288px" />
</div>

&nbsp;

```python
>>> # ;y  (yellow foreground)
>>> # ;:  (stop colors)
>>> cp.print("[ [;y]info[;:] ] program result")
```

<div align="center" title='"[ info ] program result" printed with "info" in yellow'>
  <img alt='"[ info ] program result" printed with "info" in yellow' src="/images/colorparse-example-result-2-dark.png" width="288px" />
  <img src="/images/colorparse-example-result-2-light.png" width="288px" />
</div>

&nbsp;

```python
>>> # ;R  (strong red foreground)
>>> # ;+b (start bold)
>>> # ;-b (stop bold)
>>> # :y  (yellow background)
>>> cp.print("[ /;R;+b/warning/;-b :y/!/;: ] program warning!")
```

<div align="center" title='"[ warning ! ] program warning!" printed with the first "warning" in bold, "warning !" in red and the first "!" with yellow background'>
  <img alt='"[ warning ! ] program warning!" printed with the first "warning" in bold, "warning !" in red and the first "!" with yellow background' src="/images/colorparse-example-result-3-dark.png" width="288px" />
  <img src="/images/colorparse-example-result-3-light.png" width="288px"/>
</div>
  
&nbsp;

```python
>>> # All the following strings should get parsed into the same result
>>> to_parse = [
>>>     ";r/roman empire",
>>>     "/;r/roman empire",
>>>     "[ ;r ]roman empire"
>>> ]
>>> 
>>> cp.paint(to_parse[0]) == cp.paint(to_parse[1]) == cp.paint(to_parse[2])
True
```
  
&nbsp;

### Module

When importing this module, there are the following accessible methods and variables:

#### Methods

<details>
<summary><code>print(*value, overflow:bool|None, sep:str|None, end:str|None, file:SupportsWrite[str]|None, flush:bool|None)</code></summary>

- **Description**: Prints a string with commands converted to ANSI escape sequences.

- **Examples**:

  ```python
  >>> cp.print("[", ";C/DEBUG", "]", "information...")
  ```

  ```python
  >>> test_variable = "message"
  >>> cp.print(";m;+b/warning: ", test_variable, overflow=True, sep="")
  ```

</details>
<details>
<summary><code>paint(*value, overflow:bool|None, sep:str|None, end:str|None)</code></summary>

- **Description**: Returns a string commands converted to ANSI escape sequences.

  **NOTE**: This is basically the same as `print`, but instead of directly sending output (to stdout or whatever `file` is set to) it returns the parsed string.

- **Examples**:

  ```python
  >>> cp.print("[", ";C/DEBUG", "]", "information...")
  '[\x1b[0m \x1b[38;5;45mDEBUG\x1b[0m ]\x1b[0m information...\x1b[0m'
  ```

  ```python
  >>> test_variable = "message"
  >>> cp.print(";m;+b/warning: ", test_variable, overflow=True, sep="")
  '\x1b[38;5;127m\x1b[1mwarning: message\x1b[0m'
  ```

</details>
<details>
<summary><code>register_command(name:str, command:str, value:str) -> None</code></summary>

- **Description**: Register a `Command` object. `name` is used to store the command as a variable and `command` is used when parsing (both `name` and `command` must follow Python's variable naming rules).

  **NOTE**: When adding a style command, `command` expects exactly 3 characters: ";<`+ or -`><`character`>" like ";+b" for starting bold (technically `:` can be used instead of `;`, but it doesn't affect anything).

- **Examples**:

  ```python
  >>> # A nice purple color
  >>> cp.register_command("MY_COMMAND", ";cmd", "\033[38:5:98m")
  >>> 
  >>> # Accessible from "colors.foreground.MY_COMMAND" or by its command:
  >>> cp.print(";cmd/testing new color")
  ```

  ```python
  >>> # "Framed" style. 
  >>> cp.register_command("SPECIAL_STYLE", ";+F", "\033[51m")
  >>> 
  >>> # Accessible from "styles.start.SPECIAL_STYLE" or by its command:
  >>> cp.print(";+F/probably won't show on most terminals!")
  ```

</details>
<details>
<summary><code>configuration(tag:str|_Options._OptionAttr, **values) -> None</code></summary>

- **Description**: Set global default values (or options in the case of flags) to a given value/state.

- **Examples**:

  ```python
  >>> # Set argument `overflow` to default to True for `paint`.
  >>> cp.configuration(cp.options.print, overflow=True) # "print" can also be used instead
  ```

  ```python
  >>> # Set "ignore_special" flag to be True (ignores special characters like tabs or newlines)
  >>> cp.configuration("flags", ignore_special=True)
    ```

</details>
<details>
<summary><code>show_commands()</code></summary>

- **Description**: Print a generated list of commands with their information

</details>

&nbsp;

#### Variables

<details>
<summary><code>colors</code></summary>

- **Description**: A container for both `foreground` and `background` types, and each one containing themselves the respective registered colors.
  
- **Examples**::

  ```python
  >>> cp.colors.foreground.RED
  ```

  ```python
  >>> cp.colors.background.DARK_GREEN
  ```

</details>
<details>
<summary><code>styles</code></summary>

- **Description**: A container for both `start` and `end` types, and each one containing themselves the respective registered styles.
  
- **Examples**:

  ```python
  >>> cp.styles.start.BOLD
  ```

  ```python
  >>> cp.styles.end.UNDERLINE
  ```

</details>
<details>
<summary><code>options</code></summary>

- **Description**: Stores program-wide configuration for default values and data (gets updated when calling `configuration`, and technically can be used instead of the method).
  
- **Examples**:

  ```python
  >>> if cp.options.flags.ignore_special:
  >>>     ...
  ```

  ```python
  >>> cp.options.print.overflow = True
      ```

</details>
<details>
<summary><code>RESET</code></summary>

- **Description**: Contains "`\033[0m`" which is equal to "`;.`" command (stops all commands).

</details>

&nbsp;

### Script

You can use the following to get more in-depth help for how to use colorparse in the terminal:

```terminal
colorparse --help
```

## Default commands

<div align="center">
<table>
<thead>
  <tr>
    <th colspan="2" valign="top"><b>Colors</b></br>foreground: <code>;&lt;value&gt;</code></br>background: <code>:&lt;value&gt;</code></th>
    <th colspan="2" valign="top"><b>Styles</b></br>start: <code>;+&lt;value&gt;</code></br>end: <code>;-&lt;value&gt;</code></th>
    <th colspan="2" valign="top"><b>Stopping</b></th>
  </tr>
</thead>
<tbody>
    <tr>
      <td valign="middle" align="center"><b>Value</b></td>
      <td valign="middle" align="center"><b>Attribute</b></td>
      <td valign="middle" align="center"><b>Value</b></td>
      <td valign="middle" align="center"><b>Attribute</b></td>
      <td valign="middle" align="center"><b>Command</b></td>
      <td valign="middle" align="center"><b>Targets</b></td>
    </tr>
    <tr>

CHANGELOG.md
# Change Log

\* *the prefixes [t], [m] and [d] refer to terminal-only, module-only and docs-only changes, respectively.* *

<a name="toc"></a>

- [2.0.0](https://github.com/tobimd/colorparse/blob/master/CHANGELOG.md#version-2.0.0)
- [1.1.6](https://github.com/tobimd/colorparse/blob/master/CHANGELOG.md#version-1.1.6)
- [1.1.5](https://github.com/tobimd/colorparse/blob/master/CHANGELOG.md#version-1.1.5)
- [1.1.4](https://github.com/tobimd/colorparse/blob/master/CHANGELOG.md#version-1.1.4)
- [1.1.3](https://github.com/tobimd/colorparse/blob/master/CHANGELOG.md#version-1.1.3)
- [1.1.2](https://github.com/tobimd/colorparse/blob/master/CHANGELOG.md#version-1.1.2)
- [1.1.1](https://github.com/tobimd/colorparse/blob/master/CHANGELOG.md#version-1.1.1)
- [1.1.0](https://github.com/tobimd/colorparse/blob/master/CHANGELOG.md#version-1.1.0)
- [1.0.0](https://github.com/tobimd/colorparse/blob/master/CHANGELOG.md#version-1.0.0)
- [0.0.2](https://github.com/tobimd/colorparse/blob/master/CHANGELOG.md#version-0.0.2)
- [0.0.1](https://github.com/tobimd/colorparse/blob/master/CHANGELOG.md#version-0.0.1)

<a name="latest"></a>
<a name="version-2.0.0"></a>
<a name="2.0.0"></a>

## version 2.0.0   (2022 - 08 - 06)

<a name="version-1.1.6"></a>
<a name="1.1.6"></a>

## version 1.1.6   (2020 - 03 - 12)

- Fixed color code overflow when ending specific color types consecutively (e.g. "/;g:w/ green text/;; with white background/:: would make this text green", returns the part of the string "would make this text green" with the former color green).

<a name="version-1.1.5"></a>
<a name="1.1.5"></a>

## version 1.1.5   (2020 - 03 - 12)

- Fixed raised error when using RGB or HEX color codes without true color active (now just ignores the color code).

<a name="version-1.1.4"></a>
<a name="1.1.4"></a>

## version 1.1.4   (2020 - 03 - 10)

- Re-did regular expression that obtained color codes, now works properly.
- Fixed ``end`` argument of ``paint`` function.
- Now, by default, ``end`` uses ``Color.ENDC`` before the new line character.
- Fixed the white color being the same as light gray (non-true color version).
- [d] Changed some examples and other descriptions.

<a name="version-1.1.3"></a>
<a name="1.1.3"></a>

## version 1.1.3   (2019 - 08 - 06)

- [m] Fixed ``end`` argument for the ``paint`` function not working properly.

<a name="version-1.1.2"></a>
<a name="1.1.2"></a>

## version 1.1.2   (2019 - 08 - 03)

- color codes are parsed in ``sep`` and ``end`` arguments (``-s`` and ``-e`` arguments in the terminal, respectively)

<a name="version-1.1.1"></a>
<a name="1.1.1"></a>

## version 1.1.1   (2019 - 08 - 03)

- [m] Objects that have a ``__str__`` method (e.g. iterables) can be used in ``paint``.

- [d] Changed documentation, now using rST (sphinx) instead of md (mkdocs).

<a name="version-1.1.0"></a>
<a name="1.1.0"></a>

## version 1.1.0   (2019 - 08 - 01)

- [t] Fixed problem with reading special characters on Windows.

- [t] Improved (and fixed some grammatical errors) in the help menu.
- [t] Replaced ``-r`` (``--read-special``) with ``-I`` (uppercase i) or ``--ignore-special`` (inverted roles).
- [t] Added new optional argument ``-p`` or ``--position``.
- [t] Added new optional argument ``-S`` (uppercase s) or ``--strip``.
- [m] Added more documentation for the functions insde the module (use built-in ``help`` function).
- [d] Created a documentation page.

<a name="version-1.0.0"></a>
<a name="1.0.0"></a>

## version 1.0.0   (2019 - 07 - 30)

- [t] The use of ``colorparse`` without arguments, results in the usage help being displayed.

- [t] Grouped the option arguments ``-v`` and ``-c`` as mutually exclusive arguments.
- [t] Added new optional argument for reading strings from input file(s) with ``-i`` or ``--input-file``.
- [t] Replaced the optional argument ``-f`` for ``-o`` (``--output-file``).
- [t] Changed the optional argument for ``overflow`` to ``-O`` (uppercase o) or ``--overflow``.
- [t] Added support for special characters to be read from the terminal input ``-r`` or ``--read-special``.

<a name="version-0.0.2"></a>
<a name="0.0.2"></a>

## version 0.0.2   (2019 - 07 - 29)

- Fixed Windows script file

<a name="version-0.0.1"></a>
<a name="0.0.1"></a>

## release version 0.0.1   (2019 - 07 - 29)


color.py
#!/usr/bin/python3
import os
import sys
import re
import argparse


__version__ = '1.0.0'


class _Defaults:
    paint = {
        'out': True,
        'overflow': False,
        'sep': ' ',
        'end': '\n',
        'file': sys.stdout,
        'flush': False,
    }

    _color_list = {
        'rr': 'DARK_RED',
        'oo': 'DARK_ORANGE',
        'yy': 'DARK_YELLOW',
        'gg': 'DARK_GREEN',
        'cc': 'DARK_CYAN',
        'bb': 'DARK_BLUE',
        'pp': 'DARK_PURPLE',
        'mm': 'DARK_MAGENTA',
        'r':  'RED',
        'o':  'ORANGE',
        'y':  'YELLOW',
        'g':  'GREEN',
        'c':  'CYAN',
        'b':  'BLUE',
        'p':  'PURPLE',
        'm':  'MAGENTA',
        'R':  'STRONG_RED',
        'O':  'STRONG_ORANGE',
        'Y':  'STRONG_YELLOW',
        'G':  'STRONG_GREEN',
        'C':  'STRONG_CYAN',
        'B':  'STRONG_BLUE',
        'P':  'STRONG_PURPLE',
        'M':  'STRONG_MAGENTA',
        'k':  'BLACK',
        'aa': 'DARK_GRAY',
        'a':  'GRAY',
        'A':  'LIGHT_GRAY',
        'w':  'WHITE',
    }


class Color:
    @classmethod
    def __init__(cls):
        cls.ENDC = '\033[0m'  # ;;

        cls.foreground = _Foreground(False)
        cls.background = _Background()
        cls.true_color(False)

    @classmethod
    def true_color(cls, value=None):
        if value is None:
            return cls._true_color_active
        
        # set global value for true color
        cls._true_color_active = value
        cls.foreground = _Foreground(value)


class _Foreground:
    def __init__(self, true_color):
	    # if the stdout does support rgb color escape sequences
        if true_color:
            self.DARK_RED        = '\033[38;2;130;0;0m'       # ;R
            self.DARK_ORANGE     = '\033[38;2;160;50;0m'      # ;O
            self.DARK_YELLOW     = '\033[38;2;150;120;0m'     # ;Y
            self.DARK_GREEN      = '\033[38;2;0;80;0m'        # ;G
            self.DARK_CYAN       = '\033[38;2;0;120;130m'     # ;C
            self.DARK_BLUE       = '\033[38;2;20;50;130m'     # ;B
            self.DARK_PURPLE     = '\033[38;2;90;0;150m'      # ;P
            self.DARK_MAGENTA    = '\033[38;2;130;0;100m'     # ;M

            self.RED             = '\033[38;2;180;0;0m'       # ;r
            self.ORANGE          = '\033[38;2;200;90;0m'      # ;o
            self.YELLOW          = '\033[38;2;190;170;0m'     # ;y
            self.GREEN           = '\033[38;2;0;150;0m'       # ;g
            self.CYAN            = '\033[38;2;0;190;200m'     # ;c
            self.BLUE            = '\033[38;2;0;70;255m'      # ;b
            self.PURPLE          = '\033[38;2;100;0;180m'     # ;p
            self.MAGENTA         = '\033[38;2;190;0;150m'     # ;m

            self.STRONG_RED      = '\033[38;2;255;0;0m'       # ;rr
            self.STRONG_ORANGE   = '\033[38;2;255;150;0m'     # ;oo
            self.STRONG_YELLOW   = '\033[38;2;255;255;0m'     # ;yy
            self.STRONG_GREEN    = '\033[38;2;0;255;0m'       # ;gg
            self.STRONG_CYAN     = '\033[38;2;0;255;255m'     # ;cc
            self.STRONG_BLUE     = '\033[38;2;0;20;255m'      # ;bb
            self.STRONG_PURPLE   = '\033[38;2;150;0;255m'     # ;pp
            self.STRONG_MAGENTA  = '\033[38;2;225;0;225m'     # ;mm

            self.BLACK           = '\033[38;2;0;0;0m'         # ;k
            self.DARK_GRAY       = '\033[38;2;70;70;70m'      # ;A
            self.GRAY            = '\033[38;2;130;130;130m'   # ;a
            self.LIGHT_GRAY      = '\033[38;2;185;185;185m'   # ;aa
            self.WHITE           = '\033[38;2;255;255;255m'   # ;w

	    # if the stdout does *not* support rgb color escape sequences
        else:
            self.DARK_RED        = '\033[38;5;88m'   # ;R
            self.DARK_ORANGE     = '\033[38;5;130m'  # ;O
            self.DARK_YELLOW     = '\033[38;5;142m'  # ;Y
            self.DARK_GREEN      = '\033[38;5;22m'   # ;G
            self.DARK_CYAN       = '\033[38;5;31m'   # ;C
            self.DARK_BLUE       = '\033[38;5;24m'   # ;B
            self.DARK_PURPLE     = '\033[38;5;54m'   # ;P
            self.DARK_MAGENTA    = '\033[38;5;90m'   # ;M

            self.RED             = '\033[38;5;124m'  # ;r
            self.ORANGE          = '\033[38;5;166m'  # ;o
            self.YELLOW          = '\033[38;5;184m'  # ;y
            self.GREEN           = '\033[38;5;34m'   # ;g
            self.CYAN            = '\033[38;5;39m'   # ;c
            self.BLUE            = '\033[38;5;27m'   # ;b
            self.PURPLE          = '\033[38;5;57m'   # ;p
            self.MAGENTA         = '\033[38;5;127m'  # ;m

            self.STRONG_RED      = '\033[38;5;196m'    # ;rr
            self.STRONG_ORANGE   = '\033[38;5;202m'  # ;oo
            self.STRONG_YELLOW   = '\033[38;5;226m'   # ;yy
            self.STRONG_GREEN    = '\033[38;5;82m'   # ;gg                
            self.STRONG_CYAN     = '\033[38;5;45m'   # ;cc
            self.STRONG_BLUE     = '\033[38;5;21m'   # ;bb
            self.STRONG_PURPLE   = '\033[38;5;93m'   # ;pp
            self.STRONG_MAGENTA  = '\033[38;5;200m'   # ;mm

            self.BLACK           = '\033[38;5;232m'    # ;k
            self.DARK_GRAY       = '\033[38;5;238m'  # ;A
            self.GRAY            = '\033[38;5;244m'  # ;a
            self.LIGHT_GRAY      = '\033[38;5;250m'  # ;aa
            self.WHITE           = '\033[38;5;256m'   # ;w


class _Background:
    def __init__(self):
        self.DARK_RED        = '\033[48;5;88m'   # :R
        self.DARK_ORANGE     = '\033[48;5;130m'  # :O
        self.DARK_YELLOW     = '\033[48;5;142m'  # :Y
        self.DARK_GREEN      = '\033[48;5;22m'   # :G
        self.DARK_CYAN       = '\033[48;5;31m'   # :C
        self.DARK_BLUE       = '\033[48;5;19m'    # :B
        self.DARK_PURPLE     = '\033[48;5;54m'   # :P
        self.DARK_MAGENTA    = '\033[48;5;127m'  # :M

        self.RED             = '\033[48;5;124m'  # :r
        self.ORANGE          = '\033[48;5;166m'  # :o
        self.YELLOW          = '\033[48;5;184m'  # :y
        self.GREEN           = '\033[48;5;34m'   # :g
        self.CYAN            = '\033[48;5;45m'   # :c
        self.BLUE            = '\033[48;5;27m'   # :b
        self.PURPLE          = '\033[48;5;93m'   # :p
        self.MAGENTA         = '\033[48;5;165m'  # :m

        self.STRONG_RED      = '\033[48;5;196m'    # :rr
        self.STRONG_ORANGE   = '\033[48;5;202m'  # :oo
        self.STRONG_YELLOW   = '\033[48;5;226m'   # :yy
        self.STRONG_GREEN    = '\033[48;5;82m'   # :gg
        self.STRONG_CYAN     = '\033[48;5;45m'   # :cc
        self.STRONG_BLUE     = '\033[48;5;21m'   # :bb
        self.STRONG_PURPLE   = '\033[48;5;128m'   # :pp
        self.STRONG_MAGENTA  = '\033[48;5;200m'   # :mm

        self.BLACK           = '\033[48;5;232m'    # :k
        self.DARK_GRAY       = '\033[48;5;238m'  # :A
        self.GRAY            = '\033[48;5;244m'  # :a
        self.LIGHT_GRAY      = '\033[48;5;250m'  # :aa
        self.WHITE           = '\033[48;5;255m'   # :w


def _clamp(number):
    if number == '':
        number = 0

    if type(number) == str:
        number = int(number)

    if number < 0:
        number = 0
    
    if number > 255:
        number = 255
    
    return str(number)


def _hex_to_rgb(string):
    r = int(string[:2], 16)
    g = int(string[2:4], 16)
    b = int(string[4:], 16)

    return map(_clamp, (r, g, b))


def _color_repl(matchobj):
    # if it was escaped, then return without the backslash
    if '\\' in matchobj[0]:
        return matchobj[0].replace('\\', '')

    # get the string of the matched color code and remove unwanted characters
    string = re.sub(r'\[|\]|\(|\)|/| ', '', matchobj[0])

    # if matched string is ':;' or ';:', then return ENDC (end color)
    if string == ':;' or string == ';:':
        return Color.ENDC

    col_type, col_val = string[0], string[1:]

    # custom color code
    if '#' in col_val or '=' in col_val:
        mode, val = col_val[0], col_val[1:]

        r, g, b = (map(_clamp, val.split(',')) if mode == '='
                   else _hex_to_rgb(val + '0' * (6 - len(val))))

        return f'\033[38;2;{r};{g};{b}m'

    # foreground color code
    elif col_type == ';':
        return getattr(Color.foreground, _Defaults._color_list[col_val])

    # background color code
    else:
        return getattr(Color.background, _Defaults._color_list[col_val])


def _color_format(string):
    # a failed attempt to make this a less convoluted regex
    start = r'((\[(?=[^/\)]+\]))\s*|(\((?=[^/\]]+\)))\s*|/)?('
    prefix = r'\\?(?:(;)|(:))'
    color = r'(;|:|[ROYGCBPMkAw]|([roygcbpma])\8?'
    rgb = r'|(?(5)(= ?\d{0,3}\s?,\s?\d{0,3}\s?,\s?\d{0,3}'
    hex_rgb = r'|# ?[0-9A-Fa-f]{0,6})|\5)'
    suffix = r'))(/|(?(2)\s*\]|(?(3)\s*\))))?'

    # set regex to use (with true color, use custom color codes)
    if Color._true_color_active:
        regex = f'{start}{prefix}{color}{rgb}{hex_rgb}{suffix}'
        endc_regex = f'{start}{prefix}{color}{rgb}{hex_rgb}{suffix}'

    else:
        regex = f'{start}{prefix}{color}{suffix}'
        endc_regex = f'{start}{prefix}{color}|;|:{suffix}'

    while ';;' in string or '::' in string:
        if ';::' in string or ':;;' in string:
            string = string.replace(';::', '[;:]:')
            string = string.replace(':;;', '[;:];')

        codes = re.finditer(endc_regex, string)
        saved = []
        for matchobj in codes:
            code = matchobj[0]

            if ';;' not in code and '::' not in code:
                saved.append(matchobj)

            else:
                inv_code = '::' if ';;' in code else ';;'

                for mo in saved[::-1]:
                    if inv_code[0] in mo[0]:
                        
                        string = string.replace(code, '[;:]' + mo[0], 1)
                        saved = []
                    
                if len(saved) > 0:
                    string = string.replace(code, '[;:]', 1)
                    continue

    return re.sub(regex, _color_repl, string)


def paint(*strings, **options):
    endc = '\033[0m'

    # get the options' arguments
    _out = options.get('out', _Defaults.paint['out'])
    _overflow = options.get('overflow', _Defaults.paint['overflow'])
    _sep = options.get('sep', _Defaults.paint['sep'])
    _end = options.get('end', _Defaults.paint['end'])
    _file = options.get('file', _Defaults.paint['file'])
    _flush = options.get('flush', _Defaults.paint['flush'])
    
    # if the input strings is a tuple of tuples ((strings, ...), )
    if len(strings) > 0 and type(strings[0]) != str:
        strings = strings[0]

    # if overflow is true, then color the strings as one
    if _overflow:
        result = _color_format(_sep.join(strings)) + endc
    
    # else, colors one by one and the join them
    else:        
        result = (endc + _sep).join(map(_color_format, strings)) + endc

    # if out is True, then print
    if _out:
        print(result, end=_end, file=_file, flush=_flush)

    return result


def codes():
    help_string = """
 [background] [foreground] [code]   [name]
   "(;w):(;a)<code>(;:)"   "(;w);(;a)<code>(;:)"\n
   [:rr]        [;:]     [;rr]red     [;:]     [;w]rr[;:]   (DARK_RED)
   [:oo]        [;:]     [;oo]orange  [;:]     [;w]oo[;:]   (DARK_ORANGE)
   [:yy]        [;:]     [;yy]yellow  [;:]     [;w]yy[;:]   (DARK_YELLOW)
   [:gg]        [;:]     [;gg]green   [;:]     [;w]gg[;:]   (DARK_GREEN)
   [:cc]        [;:]     [;cc]cyan    [;:]     [;w]cc[;:]   (DARK_CYAN)
   [:bb]        [;:]     [;bb]blue    [;:]     [;w]bb[;:]   (DARK_BLUE)
   [:pp]        [;:]     [;pp]purple  [;:]     [;w]pp[;:]   (DARK_PURPLE)
   [:mm]        [;:]     [;mm]magenta [;:]     [;w]mm[;:]   (DARK_MAGENTA)

   [:r ]        [;:]     [;r ]red     [;:]     [;w]r [;:]   (RED)
   [:o ]        [;:]     [;o ]orange  [;:]     [;w]o [;:]   (ORANGE)
   [:y ]        [;:]     [;y ]yellow  [;:]     [;w]y [;:]   (YELLOW)
   [:g ]        [;:]     [;g ]green   [;:]     [;w]g [;:]   (GREEN)
   [:c ]        [;:]     [;c ]cyan    [;:]     [;w]c [;:]   (CYAN)
   [:b ]        [;:]     [;b ]blue    [;:]     [;w]b [;:]   (BLUE)
   [:p ]        [;:]     [;p ]purple  [;:]     [;w]p [;:]   (PURPLE)
   [:m ]        [;:]     [;m ]magenta [;:]     [;w]m [;:]   (MAGENTA)

   [:R ]        [;:]     [;R ]red     [;:]     [;w]R [;:]   (STRONG_RED)
   [:O ]        [;:]     [;O ]orange  [;:]     [;w]O [;:]   (STRONG_ORANGE)
   [:Y ]        [;:]     [;Y ]yellow  [;:]     [;w]Y [;:]   (STRONG_YELLOW)
   [:G ]        [;:]     [;G ]green   [;:]     [;w]G [;:]   (STRONG_GREEN)
   [:C ]        [;:]     [;C ]cyan    [;:]     [;w]C [;:]   (STRONG_CYAN)
   [:B ]        [;:]     [;B ]blue    [;:]     [;w]B [;:]   (STRONG_BLUE)
   [:P ]        [;:]     [;P ]purple  [;:]     [;w]P [;:]   (STRONG_PURPLE)
   [:M ]        [;:]     [;M ]magenta [;:]     [;w]M [;:]   (STRONG_MAGENTA)

   [:k ]        [;:]    [;k ]black     [;:]    [;w]k [;:]   (BLACK)
   [:aa]        [;:]    [;aa]dark gray [;:]    [;w]aa[;:]   (DARK_GRAY)
   [:a ]        [;:]    [;a ]gray      [;:]    [;w]a [;:]   (GRAY)
   [:A ]        [;:]    [;A ]light gray[;:]    [;w]A [;:]   (LIGHT_GRAY)
   [:w ]        [;:]    [;w ]white     [;:]    [;w]w [;:]   (WHITE)

               end color   [;w]\;: | \:;[;:]   (ENDC)

"""

    extra_string = """
\t- "CUSTOM RGB":     ;=      (rgb)         [from 0 to 255, comma separated]
\t- "CUSTOM HEX":     ;#      (hexadecimal) [from 000000 to ffffff]
    """

    paint(help_string, out=True)

    if Color.true_color():
        paint(extra_string, out=True)


def change_defaults(fn, **kwargs):
    if callable(fn):
        fn = fn.__name__

    for k, v in kwargs.items():
        getattr(_Defaults, fn)[k] = v


def arg_parser():
    # initiate argument parser
    parser = argparse.ArgumentParser(prog='color (color.py)')

    # arguments
    parser.add_argument('string', 
                        help='a string that may contain color codes',
                        nargs='*',
                        default='')

    parser.add_argument('-t', '--true-color',
                        help='use of rgb values for the color escape\
                              sequences, allowing customized foreground color\
                              codes and having the color set be more accurate\
                              (warning: having this option won\'t work on all\
                              terminals as they do not all have true color).',
                        action='store_true')

    parser.add_argument('-f', '--file',
                        help='specify an output file to send the resulting\
                              formatted string. If the file exists, it will\
                              be appended to the end of said file.',
                        nargs='?',
                        type=argparse.FileType('a'),
                        default=sys.stdout)

    parser.add_argument('-o', '--overflow',
                        help='make color codes overflow to other strings if\
                              the previous one has not ended the color code.',
                        action='store_true')

    parser.add_argument('-s', '--sep',
                        help='specify what to use to separate string\
                              arguments.',
                        default=' ')

    parser.add_argument('-e', '--end',
                        help='specify what to use at the end of the resulting\
                              formatted string',
                        default='\n')

    parser.add_argument('-c', '--codes',
                        help='show the available color codes and exit.',
                        action='store_true')

    parser.add_argument('-v', '--version',
                        help='show the current version of this module and\
                              exit.',
                        action='version', version=f'%(prog)s v{__version__}')

    # return the arguments to "_main"
    return parser.parse_args()


Color()


def _main():
    # get the arguments
    args = arg_parser()
    
    # set true color to the argument's value
    Color.true_color(args.true_color)

    # if used '-c' or '--codes', the call 'codes' and exit
    if args.codes:
        codes()
        sys.exit(0)

    # print the result string
    paint(args.string, overflow=args.overflow,
        sep=args.sep, end=args.end, file=args.file)


if __name__ == '__main__':
    pass

else:
    if sys.platform[:3] == 'win':
        os.system('color')
